### PR TITLE
Make uSD Card Image for the Audio Mini

### PR DESCRIPTION
**Pull request by dack-fe**
Merging the build scripts, Dockerfile, and Jenkinsfile into dev for the Audio Mini Linux Image Builder

### CODE REVIEW

File: image_builder/audiomini/docker_build_audiomini.sh
Comment (fe-wickham) [inline image_builder/audiomini/docker_build_audiomini.sh]: Remove

File: image_builder/audiomini/Dockerfile
Comment (fe-wickham) [inline image_builder/audiomini/Dockerfile]: It'd be nice to add a comment that dosfstools is for installing `mkfs.vfat`. Otherwise I see us all forgetting why thats there in the not too distant future

File: image_builder/README.md
Comment (fe-wickham) [inline image_builder/README.md:5]: Woot! Woot! We have a solid image finally!

**Review by fe-wickham — COMMENTED** [review]
Looks good overall but it'd be nice to see the couple changes I mention in my other comments before merge

File: image_builder/audiomini/docker_build_audiomini.sh
Comment (dack-fe) [reply image_builder/audiomini/docker_build_audiomini.sh]: Resolved with 6868616

File: image_builder/audiomini/Dockerfile
Comment (dack-fe) [reply image_builder/audiomini/Dockerfile]: Resolved with 51025f7

File: image_builder/audiomini/build_audiomini_image.sh
Comment (tvannoy) [inline image_builder/audiomini/build_audiomini_image.sh:75]: Does this preserve symbolic links? I guess if the final image works, then it doesn't matter.

**Review by tvannoy — DISMISSED** [review]
Without being able to test it, it looks good to me. Glad to see the tool by robseb being used.

File: image_builder/audiomini/build_audiomini_image.sh
Comment (fe-wickham) [reply image_builder/audiomini/build_audiomini_image.sh:75]: I don't believe it does given the `-h` is described as `follow symlinks; archive and dump the files they point to` (it does the same for hardlinks as well).

I feel like presently we are using the `-h` out of necessity because I recall problems  when extracting without that flag but I would have to investigate further to know what problems we are running into without it.

Comment (tvannoy) [reply image_builder/audiomini/build_audiomini_image.sh:75]: Ah, I see. So links are just being replaced by the files. 

From my recollection, things will be messed up, and probably won't boot, if links are ignored. I recall running into issues where things like `/bin/sh` couldn't be found, leading to kernel panics. 

I believe `-p` preserves symbolic links, which would be an alternative to `-h`.

File: image_builder/audiomini/ReadMe.md
Comment (fe-tdavis) [inline image_builder/audiomini/ReadMe.md:24]: Not sure if it was an issue with my WSL, but when I ran this line I got some permission errors so I ended up running it with `sudo`

**Review by fe-tdavis — COMMENTED** [review]
I tested the full process and the image built and ran successfully although

File: image_builder/audiomini/ReadMe.md
Comment (fe-tdavis) [inline image_builder/audiomini/ReadMe.md:24]: I ran into permission errors in my WSL running line 24, so I ended up having to use `sudo` to run this command

**Comment by fe-tdavis** [comment]
> Without being able to test it

I'm going to be a bit of a stickler here...  If you are going to review our pull requests, please don't approve something you can't test.

**Comment by tvannoy** [comment]
> I'm going to be a bit of a stickler here...  If you are going to review our pull requests, please don't approve something you can't test.  

@fe-tdavis Fair enough. I will only make comments without explicit approval if I am not able to test. I think it's a good policy to make sure tests pass before merging. I am happy to keep reviewing PRs, especially when my review is requested.

**Comment by fe-tdavis** [comment]
> @fe-tdavis Fair enough. I will only make comments without explicit approval if I am not able to test. I think it's a good policy to make sure tests pass before merging. I am happy to keep reviewing PRs, especially when my review is requested.

:thumbsup:

**Comment by dack-fe** [comment]
What is the status on the symbolic links? Have they been tested? Is it an issue?

**Comment by fe-tdavis** [comment]
> What is the status on the symbolic links? Have they been tested? Is it an issue?

**Comment by fe-tdavis** [comment]
> What is the status on the symbolic links? Have they been tested? Is it an issue?

I did a quick test with the system as is and it appears to work.  I've used the `-h` flag in the past with my root filesystems and I haven't run into anything odd that I can remember...  I don't have a preference if we using the `-h` or switch to `-p` if that works.

**Review by fe-tdavis — APPROVED** [review]
Works as is, but if we change to the `-p` flag we should do further testing

**Comment by tvannoy** [comment]
I misspoke earlier. `-p` is for preserve permissions, which is default when run as superuser (we do want permissions preserved). 

I think it is preferable to keep the symlinks rather than dereferencing them, but I don't know if dereferencing causes any practical problems. Judging by reports so far, it doesn't appear to be an issue. My thought is that if symlinks are there in the original Ubuntu base archive, there is probably a reason, so we should keep the symlinks.  

I would think using the same tar arguments that we use in `build_rootfs.sh` would work correctly, though, as I don't recall having any issues with that before. 

We can verify if we are removing symlinks by running `find` with `-type l` on the original Ubuntu base directory and our modified rootfs.

File: image_builder/audiomini/ReadMe.md
Comment (dack-fe) [reply image_builder/audiomini/ReadMe.md:24]: Your use was most likely not in the Docker group

**Comment by fe-wickham** [comment]
> I think it is preferable to keep the symlinks rather than dereferencing them, but I don't know if dereferencing causes any practical problems. Judging by reports so far, it doesn't appear to be an issue. My thought is that if symlinks are there in the original Ubuntu base archive, there is probably a reason, so we should keep the symlinks.

I am confident it is preferable to preserve symlinks, hence I have created #25 but I don't believe it is worth stopping this from being merged because there isn't much reason for symlinks to break the image as built from working. What I would be concerned about is future configuration and maintenance of a running system because something may be setup currently with symlinks such that it only updates 1 file and without the symlinks it won't  update the other files.

**Review by fe-wickham — APPROVED** [review]
With the minor issues I raised addressed, this all looks great now

**Comment by tvannoy** [comment]
> > I think it is preferable to keep the symlinks rather than dereferencing them, but I don't know if dereferencing causes any practical problems. Judging by reports so far, it doesn't appear to be an issue. My thought is that if symlinks are there in the original Ubuntu base archive, there is probably a reason, so we should keep the symlinks.
> 
> 
> 
> I am confident it is preferable to preserve symlinks, hence I have created #25 but I don't believe it is worth stopping this from being merged because there isn't much reason for symlinks to break the image as built from working. What I would be concerned about is future configuration and maintenance of a running system because something may be setup currently with symlinks such that it only updates 1 file and without the symlinks it won't  update the other files.

Yes, exactly. I only see it being an issue if (most likely when) a package update updates a file (library or executable, most likely) that is supposed to be the target of a symlink. Symlinks are really common with `.so` libraries. 

I agree it isn't worth holding this PR up, as the image does work as is. Fully testing and getting symlinks preserved will take some while because the rootfs building/flashing operations are quite slow.

**Pull request merged**
